File: python/semanticabi/steps/InitStep.py
from functools import partial
from typing import List, Tuple, Dict, Callable

from semanticabi.abi.SemanticAbi import SemanticAbi
from semanticabi.abi.item.SemanticAbiItem import SemanticAbiItem, SemanticAbiEvent, DecodedResult
from semanticabi.common.ValueConverter import ValueConverter
from semanticabi.metadata.EthBlock import EthBlock
from semanticabi.metadata.EthLog import EthLog
from semanticabi.metadata.EthTraces import EthTrace
from semanticabi.metadata.EthTransaction import EthTransaction
from semanticabi.steps.AbiSchema import AbiSchema
from semanticabi.steps.Step import Step, TransformItem


class EventTransformItem(TransformItem):
    """
    A TransformItem that wraps an EthLog event
    """
    event: EthLog

    def __init__(self, event: EthLog, decoded_result_fn: Callable[[], DecodedResult]):
        super().__init__(decoded_result_fn)
        self.event = event

    @property
    def contract_address(self) -> str:
        return self.event['address']

    @property
    def internal_index(self) -> str:
        return str(ValueConverter.hex_to_int(self.event['logIndex']))

    @property
    def item_type(self) -> str:
        return 'event'


class FunctionTransformItem(TransformItem):
    """
    A TransformItem that wraps an EthTrace function call
    """
    function: EthTrace

    def __init__(self, function: EthTrace, decoded_result_fn: Callable[[], DecodedResult]):
        super().__init__(decoded_result_fn)
        self.function = function

    @property
    def contract_address(self) -> str:
        return self.function.to_address

    @property
    def internal_index(self) -> str:
        return self.function.trace_hash

    @property
    def item_type(self) -> str:
        return 'function'


class InitStep(Step):
    """
    A no-op step that gets initialized with the particular ABI item that it'll handle, and filters the logs
    or traces by the signature of that item, optionally also filtering those by any contract addresses specified in the
    ABI, adding an empty row for each matching log or trace
    """

    _semantic_abi: SemanticAbi
    _semantic_abi_item: SemanticAbiItem
    _schema: AbiSchema

    def __init__(self, abi: SemanticAbi, abi_item: SemanticAbiItem):
        self._semantic_abi = abi
        self._semantic_abi_item = abi_item
        self._schema = AbiSchema()

    @property
    def _abi(self) -> SemanticAbi:
        return self._semantic_abi

    @property
    def _abi_item(self) -> SemanticAbiItem:
        return self._semantic_abi_item

    @property
    def schema(self) -> AbiSchema:
        return self._schema

    def _inner_transform(self, block: EthBlock, transaction: EthTransaction) -> List[Tuple[TransformItem, List[Dict[str, any]]]]:
        results = []
        if isinstance(self._abi_item, SemanticAbiEvent):
            logs: List[EthLog] = transaction.logs_by_topic.get(self._abi_item.raw_item.hash, [])
            for log in logs:
                transform_item: EventTransformItem = EventTransformItem(log, partial(self._abi_item.decode, log))
                if self._abi.should_consider(transform_item.contract_address):
                    results.append((transform_item, [{}]))
        else:
            traces: List[EthTrace] = transaction.traces_by_topic.get(self._abi_item.raw_item.hash, [])
            for trace in traces:
                transform_item: FunctionTransformItem = FunctionTransformItem(trace, partial(self._abi_item.decode, trace))
                if self._abi.should_consider(transform_item.contract_address):
                    results.append((transform_item, [{}]))

        return results

File: js/src/semanticabi/transform/steps/InitStep.ts
import {TransformItem} from "semanticabi/transform/steps/TransformItem";
import {EthLog} from "semanticabi/metadata/EthLog";
import {DecodedResult, SemanticAbiItem} from "semanticabi/abi/item/semantic/SemanticAbiItem";
import {EthTrace} from "semanticabi/metadata/EthTrace";
import {HexToNumber} from "semanticabi/transform/column/HexToNumber";
import {Step} from "semanticabi/transform/steps/Step";
import {SemanticAbi} from "semanticabi/abi/SemanticAbi";
import {JsonObject} from "common/CommonTypes";
import {EthTransaction} from "semanticabi/metadata/EthTransaction";
import {EthBlock} from "semanticabi/metadata/EthBlock";
import {AbiSchema} from "semanticabi/transform/steps/AbiSchema";
import {SemanticAbiEvent} from "semanticabi/abi/item/semantic/SemanticAbiEvent";


/**
 * A no-op step that gets initialized with the particular ABI item that it'll handle, and filters the logs or traces by
 * the signature of that item, optionally also filtering those by any contract addresses specified in the ABI, adding an
 * empty row for each matching log or trace
 */
export class InitStep extends Step {

    public readonly schema: AbiSchema;

    constructor(
        public readonly abi: SemanticAbi,
        public readonly abiItem: SemanticAbiItem
    ) {
        super();
        this.schema = new AbiSchema([]);
    }

    get shouldTransform(): boolean {
        return true;
    }

    innerTransform(block: EthBlock, transaction: EthTransaction): [TransformItem, JsonObject[]][] {
        const transformItems = (): TransformItem[] => {
            if (this.abiItem instanceof SemanticAbiEvent) {
                const logs: EthLog[] = transaction.logsByTopic.get(this.abiItem.rawItem.hash) ?? [];
                return logs.map(log => new EventTransformItem(log, () => this.abiItem.decode(log)));
            } else {
                const traces: EthTrace[] = transaction.tracesByTopic.get(this.abiItem.rawItem.hash) ?? [];
                return traces.map(trace => new FunctionTransformItem(trace, () => this.abiItem.decode(trace)));
            }
        };

        return transformItems()
            .filter(transformItem => this.abi.shouldConsider(transformItem.contractAddress))
            .map(transformItem => [transformItem, [{}]]);
    }

}

export class EventTransformItem extends TransformItem {

    constructor(
        public readonly event: EthLog,
        decodedResultFn: () => DecodedResult
    ) {
        super(decodedResultFn);
    }

    get contractAddress(): string {
        return this.event.address;
    }

    get internalIndex(): string {
        return HexToNumber.convert(this.event.logIndex).toString();
    }

    get itemType(): string {
        return 'event';
    }
}

export class FunctionTransformItem extends TransformItem {

    constructor(
        public readonly trace: EthTrace,
        decodedResultFn: () => DecodedResult
    ) {
        super(decodedResultFn);
    }

    get contractAddress(): string {
        return this.trace.toAddress;
    }

    get internalIndex(): string {
        return this.trace.traceHash;
    }

    get itemType(): string {
        return 'function';
    }

}